pipeline/pipeline_runner.go

```diff
@@ -386,3 +399,123 @@ func Run(config *PipelineConfig) {
 
 	LogInfo.Println("Shutdown complete.")
 }
+
+func sandboxSaveAndSigquit(config *PipelineConfig) {
+	// This should only be run when the router isn't processing messages, so we
+	// try to inject a new message and exit if successful. Far from perfect,
+	// but protects in most cases.
+	var pack *PipelinePack
+	doSave := false
+
+	select {
+	case pack = <-config.injectRecycleChan:
+	case <-time.After(1 * time.Second):
+	}
+
+	if pack == nil {
+		select {
+		case pack = <-config.inputRecycleChan:
+		case <-time.After(1 * time.Second):
+		}
+	}
+
+	if pack == nil {
+		doSave = true
+	} else {
+		pack.Message.SetType("heka.router-test")
+		select {
+		case config.router.inChan <- pack:
+		case <-time.After(1 * time.Second):
+			doSave = true
+		}
+	}
+
+	if !doSave {
+		LogError.Println("Can't save sandboxes while router is processing messages.")
+		return
+	}
+
+	// Generate a report before exiting for debugging purposes.
+	config.allReportsStdout()
+
+	// If we got this far the router is presumably wedged and we'll perform the
+	// operation.  First we close the abort channel to allow any hung
+	// inject_message calls to return, then we go through and stop and destroy
+	// all of the sandboxes. Unfortunately the data structures we use to track
+	// the sandbox plugins of each type are just different enough that it's
+	// hard to write generic code, so we loop through each plugin type
+	// separately.
+	close(config.Globals.abortChan)
+	time.Sleep(1 * time.Second)
```

Comment:
Arbitrary 'waits' for stuff to finish in a multi threaded environment will eventually hit a race condition.  If we cannot just start destroying we probably need a better solution.